tracing-attributes/src/lib.rs

```diff
@@ -335,6 +451,34 @@ fn gen_body(
         ..
     } = sig;
 
+    // If we're in an async block, we'll need to do a bit of a dance in order to
+    // ensure that return type inference works properly. This is because rust is
+    // not able to infer the correct coercion when faced with e.g.:
+    // ```
+    // #[instrument]
+    // async fn foo() -> Box<FnOnce()->() + Send> {
+    //   Box::new(|| ())
+    // }
+    // ```
+    //
+    // becoming
+    //
+    // ```
+    // #[instrument]
+    // async fn foo() -> Box<FnOnce()->() + Send> {
+    //   instrument(async move { Box::new(|| ())} ).await
+    // }
+    // ```
+    //
+    // To work around this we must instead declare the inner body as an actual function
+    // (an inner function to avoid name clashes) in order to attach the proper return type
+    // to the function.
```

Comment:
Thanks for documenting what we're trying to handle here and why we do it this way, this is very helpful!